cql3/restrictions/statement_restrictions.cc

```diff
@@ -969,5 +975,229 @@ bool single_column_restriction::LIKE::is_satisfied_by(bytes_view data, const que
+            const std::vector<bytes_opt>& other_columns, const query_options& options) {
+    if (!opr.op.is_slice()) { // For EQ or NEQ, use equals().
+        throw std::logic_error("limits() called on non-slice op");
+    }
+    const auto& columns = std::get<idents>(opr.lhs);
+    if (auto multi = dynamic_pointer_cast<multi_item_terminal>(opr.rhs)) {
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            const auto cmp = columns[i]->type->as_tri_comparator()(
+                    *get_value(columns[i], selection, partition_key, clustering_key, other_columns),
+                    *rhs[i]);
+            // If the components aren't equal, then we just learned the LHS/RHS order.
+            if (cmp < 0) {
+                if (opr.op == operator_type::LT || opr.op == operator_type::LTE) {
+                    return true;
+                } else if (opr.op == operator_type::GT || opr.op == operator_type::GTE) {
+                    return false;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            } else if (cmp > 0) {
+                if (opr.op == operator_type::LT || opr.op == operator_type::LTE) {
+                    return false;
+                } else if (opr.op == operator_type::GT || opr.op == operator_type::GTE) {
+                    return true;
+                } else {
+                    throw std::logic_error("Unknown slice operator");
+                }
+            }
+            // Otherwise, we don't know the LHS/RHS order, so check the next component.
+        }
+        // Getting here means LHS == RHS.
+        return opr.op == operator_type::LTE || opr.op == operator_type::GTE;
+    } else {
+        if (columns.size() != 1) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        auto lhs = get_value(columns[0], selection, partition_key, clustering_key, other_columns);
+        if (!lhs) {
+            return false;
+        }
+        auto rhs = to_bytes_opt(opr.rhs->bind_and_get(options));
+        if (!rhs) {
+            return false;
+        }
+        return limits(*lhs, opr.op, *rhs, *columns[0]->type);
+    }
+}
+
+/// Fetches the next cell value from iter and returns its value as bytes_opt if the cell is atomic;
+/// otherwise, returns nullopt.
+bytes_opt next_value(query::result_row_view::iterator_type& iter, const column_definition* cdef) {
+    if (cdef->type->is_multi_cell()) {
+        iter.next_collection_cell();
+    } else {
+        auto cell = iter.next_atomic_cell();
+        if (cell) {
+            return cell->value().with_linearized([] (bytes_view data) {
+                return bytes(data.cbegin(), data.cend());
```

Comment:
The lineaization and conversion to bytes is sad, but I guess forced by the context.

Reply:
Yes, this is copied from existing code.